Tests/PackageLoading/ConventionTests.swift

```diff
@@ -94,3 +127,163 @@ private func fixture(files: [RelativePath], file: StaticString = #file, line: UI
+    let pkg: PackageDescription.Package
+    if let package = package {
+        pkg = package
+    } else {
+        pkg = Package(name: path.basename)
+    }
+    let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: pkg, products: [], version: nil)
+    let builder = PackageBuilder(manifest: manifest, path: path, fileSystem: fs)
+    return try builder.construct(includingTestModules: true)
+}
+
+extension PackageModel.Package {
+    var allModules: [Module] {
+        return modules + testModules
+    }
+}
+
+final class PackageBuilderTester {
+    private enum Result {
+        case package(PackageModel.Package)
+        case error(String)
+    }
+
+    /// Contains the result produced by PackageBuilder.
+    private let result: Result
+
+    /// Contains the diagnostics which have not been checked yet.
+    private var uncheckedDiagnostics = Set<String>()
+
+    /// Setting this to true will disable checking for any unchecked diagnostics prodcuted by PackageBuilder during loading process.
+    var ignoreDiagnostics: Bool = false
+
+    /// Contains the modules which have not been checked yet.
+    private var uncheckedModules = Set<Module>()
+
+    /// Setting this to true will disable checking for any unchecked module.
+    var ignoreOtherModules: Bool = false
+
+    @discardableResult
+    init(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil, _ body: @noescape (PackageBuilderTester) -> Void) {
+        do {
+            let loadedPackage = try loadPackage(path, in: fs, package: package)
+            result = .package(loadedPackage)
+            uncheckedModules = Set(loadedPackage.allModules)
+        } catch {
+            let errorStr = String(error)
+            result = .error(errorStr)
+            // FIXME: We need to extend dianostics to load warnings too.
+            // Right now warnings are dumped directly to stdout, we can use OutputByteStream
+            // instead to write warnings to a stream and then capture them for testing.
+            uncheckedDiagnostics.insert(errorStr)
+        }
+        body(self)
+        validateDiagnostics()
+        validateCheckedModules()
+    }
+
+    private func validateDiagnostics() {
+        guard !ignoreDiagnostics && !uncheckedDiagnostics.isEmpty else { return }
+        XCTFail("Unchecked diagnostics: \(uncheckedDiagnostics)")
+    }
+
+    private func validateCheckedModules() {
+        guard !ignoreOtherModules && !uncheckedModules.isEmpty else { return }
+        XCTFail("Unchecked modules: \(uncheckedModules)")
+    }
+
+    func checkError(_ str: String) {
+        switch result {
+        case .package(let package):
+            XCTFail("\(package) did not have any error")
+        case .error(let error):
+            uncheckedDiagnostics.remove(error)
+            XCTAssertEqual(error, str)
+        }
+    }
+
+    func checkModule(_ name: String, _ body: (@noescape (ModuleResult) -> Void)? = nil) {
```

Comment:
It would be nice for these functions to use the `#file` and `#line` forwarding trick we use in other test functions, so that failures show up inside the test that failure.